yt_dlp/networking/common.py

```diff
@@ -0,0 +1,431 @@
+    remove_start
+)
+
+if typing.TYPE_CHECKING:
+    from ..YoutubeDL import YoutubeDL
+
+
+class Request:
+    """
+    Represents a request to be made.
+    Partially backwards-compatible with urllib.request.Request.
+
+    @param url: url to send. Will be sanitized and auth will be extracted as basic auth if present.
+    @param data: payload data to send.
+    @param headers: headers to send.
+    @param proxies: proxy dict mapping of proto:proxy to use for the request and any redirects.
+    @param query: URL query parameters to update the url with.
+    @param method: HTTP method to use. If no method specified, will use POST if payload data is present else GET
+    @param compression: whether to include content-encoding header on request.
+    @param timeout: socket timeout value for this request.
+    """
+    def __init__(
+            self,
+            url: str,
+            data=None,
+            headers: typing.Mapping = None,
+            proxies: dict = None,
+            query: dict = None,
+            method: str = None,
+            compression: bool = True,
+            timeout: Union[float, int] = None):
+
+        url, basic_auth_header = extract_basic_auth(escape_url(sanitize_url(url)))
+
+        if query:
+            url = update_url_query(url, query)
+        # rely on urllib Request's url parsing
+        self.__request_store = urllib.request.Request(url)
+        self.method = method
+        self._headers = CaseInsensitiveDict(headers)
+        self._data = None
+        self.data = data
+        self.timeout = timeout
+
+        if basic_auth_header:
+            self.headers['Authorization'] = basic_auth_header
+
+        self.proxies = proxies or {}
+        self.compression = compression
+
+    @property
+    def url(self):
+        return self.__request_store.full_url
+
+    @url.setter
+    def url(self, url):
+        self.__request_store.full_url = url
+
+    @property
+    def data(self):
+        return self._data
+
+    @data.setter
+    def data(self, data):
+        # https://docs.python.org/3/library/urllib.request.html#urllib.request.Request.data
+        if data != self._data:
+            self._data = data
+            if 'content-length' in self.headers:
+                del self.headers['content-length']
+
+    @property
+    def headers(self) -> CaseInsensitiveDict:
+        return self._headers
+
+    @headers.setter
+    def headers(self, new_headers: CaseInsensitiveDict):
+        if not isinstance(new_headers, CaseInsensitiveDict):
+            raise TypeError('headers must be a CaseInsensitiveDict')
+        self._headers = new_headers
+
+    @property
+    def method(self):
+        return self.__method or ('POST' if self.data is not None else 'GET')
+
+    @method.setter
+    def method(self, method: str):
+        self.__method = method
+
+    def copy(self):
+        return type(self)(
+            url=self.url, data=self.data, headers=self.headers.copy(), timeout=self.timeout,
```

Comment:
Hm. It depends what data is. Pretty sure it should be bytes as I think thats all urllib accepts, but will look into it

Reply:
ah, u r right. I thought it allowed dicts 😳

Reply:
Yeah data is either bytes or file like object

https://docs.python.org/3/library/urllib.request.html#urllib.request.Request

Requests accepts dict, which it will convert to json. But I think we should limit what is accepted to prevent people from accidentally trying. https://requests.readthedocs.io/en/latest/api/#requests.Session.request 


Would having a `json` parameter be any use?

Reply:
> Requests accepts dict, which it will convert to json. But I think we should limit what is accepted to prevent people from accidentally trying. [requests.readthedocs.io/en/latest/api/#requests.Session.request](https://requests.readthedocs.io/en/latest/api/#requests.Session.request)

Agreed

> Would having a json parameter be any use?

Not worth it imo